README.md
# Cucumber Test

Made easy - Cucumber testing using webdriver.io, instead of writing complicated
test code that only developers can understand, Cucumber maps an ordinary
language to code and allows to start with the test process in the early stages
of your product development. `cucumber-test` is a fork of
`cucumber-boilerplate`.

## Quick start

1. npm install -g cucumber-test selenium-standalone
2. selenium-standalone install
3. selenium-standalone start
4. create folder `features`
5. add file `features/simple.feature`
6. run `cct` (or `cct -b chrome,firefox,safari --timeout 40000 -i 3`)

if step 2,3 show some warnings, might be behind proxy / VPN, please check section `Behind Proxy / VPN:` below.

```cucumber
@simple
Feature:
    Search on Google

Scenario: Navigate to Google
    Given I open the url "https://www.google.com"
    Then I expect that the title is "Google"

Scenario: Search cucumber-test
    When I set "cucumber-test" to the inputfield "[name=q]"
    And I expect that element "input.lsb" becomes visible
    And I click on the button "input.lsb"
    Then I expect that element "a[href='https://cucumber.io/']" becomes visible
```

## Extended tags
Extended tags will simplify how to write cucumber with different browser & environment, those information are available on the browser object, utilize during execution of the cucumber file:
```js
// Browsers
@__chrome            ><  @__non_chrome
@__safari            ><  @__non_safari
@__firefox           ><  @__non_firefox
@__microsoftedge     ><  @__non_microsoftedge
@__internetexplorer  ><  @__non_internetexplorer

// Mobile specific
@__android           ><  @__non_android
@__mobile            ><  @__non_mobile
@__ios               ><  @__non_ios
```
Sample usage can be seen on the example below `Search on Google`, when it get executed on the desktop browser like chrome, firefox or IE, it will pickup scenario with:

`@__non_safari @__non_mobile` and the `@__mobile` will be filtered

and viceversa will happened when it get executed on mobile.

## Appium
### Android Device
please check the developer tools security options, ensure them to be checked.
* USB debugging
* USB debugging(Security settings)

```cucumber
@simple
Feature: Search on Google
    Search should be on the google website
    and the first should be cucumber.io

Scenario: Navigate to Google
    Given I open the url "https://www.google.com"
    Then I expect that the title is "Google"

@__non_safari @__non_mobile
Scenario: Search cucumber-test on desktop browser
    When I set "cucumber-test" to the inputfield "[name=q]"
    And I expect that element "input.lsb" becomes visible
    And I click on the button "input.lsb"
    Then I expect that element "a[href='https://cucumber.io/']" becomes visible

@__mobile
Scenario: Search cucumber-test on mobile browser
    When I set "cucumber-test" to the inputfield "[name=q]"
    And I expect that element "[name=btnG]" becomes visible
    And I click on the button "[name=btnG]"
    Then I expect that element "a[href='https://cucumber.io/']" becomes visible
```
```bash
# Install appium server
npm install -g appium appium-doctor
# run appium & connect your android device using USB
appium
# check deviceName - adb devices & pass to deviceName:android version
cct --android f344ee26:7.0
```

## Config & Variables
Setting up browsers capabilities and adding variables that can be use inside steps statement. Put `config.js` in the current folder (above features folder) where you will run cucumber-test.

`share.feature`
```cucumber
Feature: Share Scenario
    To be share during automation

Scenario: Navigate to Google
    Given I open the url "https://www.google.com"
    Then I expect that the title is "Google"

@__non_mobile @__chrome @__non_url:localhost
Scenario: Search cucumber-test
    When I type "cucumber-test" to the inputfield "${g.q}"
    And I expect that element "${g.btnG}" becomes visible
    And I click on the button "${g.btnG}"
    Then I expect that element "a[href='https://cucumber.io/']" becomes visible

@__non_mobile @__non_chrome
Scenario: Search cucumber-test
    When I type "cucumber-test *" to the inputfield "${g.q}"
    And I expect that element "${g.btnG}" becomes visible
    And I click on the button "${g.btnG}"
    Then I expect that element "a[href='https://en.wikipedia.org/wiki/Cucumber_(software)']" becomes visible

@__mobile
Scenario: Search cucumber-test
    When I type "cucumber-test mobile" to the inputfield "${g.q}"
    And I expect that element "${g.btnM}" becomes visible
    And I click on the button "${g.btnM}"
    Then I expect that element "a[href='https://cucumber.io/']" becomes visible
```

`google/google.feature`
```cucumber
Feature: Search on Google
    Search should be on the google website
    and the first should be cucumber.io

Scenario: Navigate to Google
    Given ...

Scenario: Search cucumber-test
    When ...
```

`config.js`
```js
module.exports = (faker) => {
    return {
        browsers: {
            chrome: {},
            firefox: {},
            safari: { platform: 'MAC' },
            edge: { browserName: 'microsoftEdge' },
            ie: { browserName: 'internet explorer',
                version: '11'
            }
        },
        browserstack: {
            browsers: {
                edge: {
                    unhandledPromptBehavior: 'accept',
                    platform: 'WINDOWS',
                    version: '16'
                },
            },
            browser: 'edge',
            timeout: 50000,
            retry: 7
        },
        saucelabs: {
            browsers: {
                chrome: {
                    platform: 'WIN7',
                    version: '65'
                },
                firefox: {
                    platform: 'WIN7',
                    version: '59'
                },
                "firefox:m52": {
                    platform: 'MAC',
                    version: '52'
                },
                safari: {
                    platform: 'macOS 10.12',
                    version: '10.1'
                },
                edge: {
                    platform: 'Windows 10',
                    version: '16'
                },
                ie: {
                    platform: 'WIN7'
                },
                ios: {
                    appiumVersion: '1.7.2',
                    deviceName:"iPhone 8 Plus Simulator",
                    deviceOrientation: 'portrait',
                    platformName: 'iOS',
                    platformVersion: '11.2',
                    browserName: 'Safari',
                },
                android: {
                    appiumVersion: '1.7.2',
                    deviceName:"Android Emulator",
                    platformName: 'Android',
                    platformVersion: '6.0',
                    browserName: 'Chrome',
                },
            },
            browser: 'chrome,firefox,safari,edge,ie',
            timeout: 50000,
            retry: 7
        },
        perfecto: {},
        remote: {
            remote: 'http://localhost:4444',
            browser: 'firefox,chrome',
            timeout: 40000,
            retry: 6
        },
        base: {
            name: 'CCT',
            cloud: 'saucelabs',
            android: 'f344ee26:7.0',
            browser: 'firefox,chrome,safari',
            inputPause: 500,
            timeout: 20000,
            instances: 1,
            retry: 3
        },
        tags: {
            __test: function(vars, string) {
                vars.g.search = `cucumber-io`;
                return true;
            }
        },
        vars: {
            g: {
                q: '[name=q]',
                btnG: 'input.lsb',
                btnM: 'button._S6q',
                search: 'cucumber-test',
                url: 'https://google.com',
                cucumberIo: "a[href='https://cucumber.io/']",
                email: faker.internet.email()
            }
        }
    }
}
```
`config.js` will automatically pickup by runner or you can provide the fullpath in CLI:
```bash
cct -f google --config ./config.js -b chrome
cct -f google -b chrome,safari,firefox
```

## Integration with: BrowserStack
Add these env variables from your BrowserStack user & key:
```bash
export BROWSERSTACK_ACCESS_KEY=xxXXXxXxXxXxxXxXXXxx
export BROWSERSTACK_USERNAME=cucumber-test
cct -c browserstack
```

![alt #D Saucelabs](docs/BrowserStack.gif)

## Integration with: Sauce Labs
Add these env variables from your SauceLabs user & key:
```bash
export SAUCE_ACCESS_KEY=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
export SAUCE_USERNAME=cucumber-test
cct -c saucelabs:connect
cct -c saucelabs
```

![alt #D Saucelabs](docs/saucelabs.png)

## Behind Proxy / VPN:

```bash
export NODE_TLS_REJECT_UNAUTHORIZED=0
npm install -g selenium-standalone@latest
selenium-standalone install
selenium-standalone start
cct -t '@smoke'
```

## Dev Mode

```bash
cd ~/.nvm/versions/node/v8.8.1/lib/node_modules
ln -s /Users/wharsojo/Dev/cucumber-test cucumber-test
```

## License

MIT


CHANGELOG.md
# Changelog

## [1.2.0] - 2018-03-28

### Share Scenario with Tags
share.feature now support multiple same scenario with different tags

```cucumber
@__chrome
Scenario: Search cucumber-test
    When I type "cucumber-test chrome" to the inputfield "${g.q}"

@__non_chrome
Scenario: Search cucumber-test
    When I type "cucumber-test non chrome" to the inputfield "${g.q}"
```
```cucumber
# those scenarios will be use as one scenario
Scenario: Search cucumber-test
    When ...
```

### config.js: `general` change to `base`
change the keyword of `general` to `base`. `base` more suitable for the base config.

## [1.1.0] - 2018-03-07

### Share Scenario
Repetitive scenario can be store to share.feature

## [1.0.17] - 2018-02-14

### Add several options:
* "--uaIphone" spoofing user agent as iPhone under chrome browser
* "--uaGalaxy" spoofing user agent as Samsung Galaxy under chrome browser
* "--timeout" default timeout can be overwrite [20000ms]

## [1.0.8] - 2018-02-12

### Published

* Published to npm


src/cct.js
const fs = require('fs');
const url = require('url');
const faker = require('faker');
const program = require('commander');
const webdriverio = require('webdriverio');
const { Launcher } = require('webdriverio');
const _merge = require('lodash/merge');

program.version('1.2.7');
program.option('-f, --features [path]', 'location of features/[path]');
program.option('-t, --tags [tags]', 'run features filtered by tags');
program.option('-r, --remote [host]', 'remote server [http://ex.com:4444]');
program.option('-c, --cloud [provider]', 'cloud [saucelabs, browserstack, perfecto]');
program.option('-i, --instances [instances]', 'max instances');
program.option('-b, --browser [browser]', 'target browser');
program.option('-n, --name [name]', 'capability name');
program.option('--timeout [timeout]', 'timeout [20000]');
program.option('--retry [retry]', 'connection retry [3]');
program.option('--config [fpath]', 'config [./config.js]');
program.option('--android [android]', 'run on android device');
program.option('--uaIphone', 'chrome w/ user agent of iPhone');
program.option('--uaGalaxy', 'chrome w/ user agent of Samsung Galaxy');
program.option('--vars [json]', `vars '{"g":{"search":"automation"}}'`);

program.parse(process.argv);

if (program.config===undefined && fs.existsSync(process.cwd()+'/config.js')) {
    program.config = 'config.js';
}

let cpath;
let config = {};
if (program.config) {
    cpath = process.cwd()+'/'+program.config;
    config = require(cpath)(webdriverio, faker);
}

let _originalTags = 'not @Pending';
if (program.tags) {
    _originalTags = `${program.tags} and (not @Pending)`;
}

let specs = ['./features/**/*.feature'];
if (program.features) {
    specs = [`./features/${program.features}/**/*.feature`];
}

let options = {services: []}; // services: ['firefox-profile'],
let base = config.base || {};
let browsers = config.browsers || {};
let remoteConfig = {};

let browser = base.browser || 'chrome';
let timeout = base.timeout || 20000;
let retry = base.retry || 3;

if (program.remote) {
    remoteConfig = config.remote || {};
}

console.log('Loading...');
if (program.cloud) {
    let cloud = base.cloud;
    if (program.cloud===true) {
        if (typeof(base.cloud)!=='string') {
            console.log(`option --cloud need to have params or config.js {base: {cloud:'cloudprovider'}} set`);
            process.exit(1);
        }
    } else {
        cloud = program.cloud;
    }
    const clouds = cloud.split(':');
    const provider = clouds[0];

    console.log(`Run from ${cloud}`);
    if (provider==='saucelabs') {
        options.services.push('sauce');
        if (clouds[1]==='connect') {
            options.sauceConnect = true;
        }
        options.user = process.env.SAUCE_USERNAME;
        options.key = process.env.SAUCE_ACCESS_KEY;
    } else if (provider==='browserstack') {
        options.services.push('browserstack');
        options.user = process.env.BROWSERSTACK_USERNAME;
        options.key = process.env.BROWSERSTACK_ACCESS_KEY;
        options.browserstackLocal = true;
    }
    if (config[provider]) {
        remoteConfig = _merge(remoteConfig, config[provider]);;
    }
}

if (typeof(program.remote)==='string') {
    remoteConfig.remote = program.remote;
}

if (remoteConfig.remote) {
    console.log('Remote>>>:', remoteConfig.remote);
    const myURL = url.parse(remoteConfig.remote);
    options.protocol = myURL.protocol.replace(':', '') || 'http';
    options.port = myURL.port || 4444;
    options.host = myURL.hostname;
}

if (remoteConfig.browsers) {
    browsers = _merge(browsers, remoteConfig.browsers);
}

if (remoteConfig.browser) {
    browser = remoteConfig.browser;
}

if (remoteConfig.timeout) {
    timeout = remoteConfig.timeout;
}

if (remoteConfig.retry) {
    retry = remoteConfig.retry;
}

if (program.browser) {
    browser = program.browser;
}

if (program.timeout) {
    timeout = program.timeout;
}

if (program.retry) {
    retry = program.retry;
}

if (typeof(program.name)==='string') {
    base.name = program.name;
}

let browserIds = browser.split(',');
if (!base.instances) {
    base.instances = browserIds.length;
}

if (program.instances) {
    base.instances = program.instances;
}

options = _merge(options, {
    connectionRetryCount: retry,
    waitforTimeout: timeout - 10000,
    // firefoxProfile: {"security.tls.version.max": 1},
    maxInstances: base.instances,
    cucumberOpts: {
        _originalTags,
        timeout
    },
    vars: {},
    specs,
    cpath,
    base,
});
console.log('Timeout/Retry/I:', `${timeout}/${retry}/${base.instances}`);

if (program.android) {
    if (program.android===true) {
        if (base.android) {
            program.android = base.android;
        } else {
            console.log(`
deviceName & platformVersion are required!
cct --android [deviceName:platformVersion]
            `);
            process.exit(0);
        }
    }
    const android = program.android.split(':');
    browser = 'chrome';
    options.port = '4723';
    options.services.push('appium');
    options.maxInstances = 1;
    options.capabilities = [{
        name: base.name,
        browserName: browser,  // browser name is empty for native apps
        appiumVersion: '1.7.2', // Appium module version
        // https://github.com/appium/appium/issues/8651
        platformName: 'Android',
        platformVersion: android[1] || '7.0',  // Android version
        deviceName: android[0], // device name of the mobile device
        newCommandTimeout: 30 * 60000,
        waitforTimeout: 10000,
        commandTimeout: 7200,
    }];
} else {
    if (browser.match(/\:[a-zA-Z]+\d+/) && program.config===undefined) {
        console.log(`Browser: ${browser} need to have option --config`);
        process.exit(1);
    }
    options.services.push('selenium-standalone');  // 'firefox-profile'
    options.capabilities = browserIds.map(bName => {
        const browserCfg = bName.split(':');
        const browserName = browserCfg[0];
        const bconfig = {
            name: base.name,
            maxInstances: 5,
            browserName,
            acceptInsecureCerts: true,
        };
        if (browserCfg[1]) {
            bconfig.version = browserCfg[1];
        };
        if (browserName === 'chrome') {
        	const args = ['disable-web-security'];
	        bconfig.chromeOptions = {args}
            if (program.uaIphone) {
                args.push('use-mobile-user-agent', 'user-agent=Mozilla/5.0 (iPhone; CPU iPhone OS 5_0 like Mac OS X) AppleWebKit/534.46 (KHTML, like Gecko) Version/5.1 Mobile/9A334 Safari/7534.48.3')
            } else if (program.uaGalaxy) {
                args.push('use-mobile-user-agent', 'Mozilla/5.0 (Linux; Android 7.0;SAMSUNG SM-G955F Build/NRD90M) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/5.2 Chrome/51.0.2704.106 Mobile Safari/537.36')
            }
        }
        if (browserName==='safari') {
            bconfig['safari.options'] = {
                technologyPreview: true,
            }
        }
        return bconfig;
    });
}

console.log('Browser:', browser,
program.uaIphone ? '--uaIphone' : '',
program.uaGalaxy ? '--uaGalaxy' : '',
program.android  ? '--android' : '',
);

if (program.config) {
    options.capabilities.forEach((obj, idx) => {
        const name = obj.browserName, version = obj.version;
        if (browsers[`${name}:${version}`]) {
            options.capabilities[idx] = Object.assign({}, options.capabilities[idx], browsers[`${name}:${version}`]);
            console.log(options.capabilities[idx]);
        } else if (browsers[name]) {
            options.capabilities[idx] = Object.assign({}, options.capabilities[idx], browsers[name]);
            console.log(options.capabilities[idx]);
        }
    });
}

if (program.vars) {
    const paramVars = JSON.parse(program.vars);
    options.vars = _merge(options.vars, paramVars);
}

const tagExpression = _originalTags;
options.cucumberOpts.tagExpression = tagExpression;
console.log('Tags:', tagExpression);

const wdio = new Launcher(`${__dirname}/../wdio.conf.js`, options);
console.log('start...');
try {
    wdio.run().then(
        code => {
            console.log('Normal exit:', code);
            process.exit(code);
        },
        error => {
            console.error('Launcher failed to start the test', error.stacktrace);
            process.exit(1);
        }
    );
} catch(e) {
    console.log('Ultimate error catch:', e);
    process.exit(1);
}
